### PR TITLE
Add Lucene segment-level fields stats

### PR DESCRIPTION
**Pull request by dnhatn**
This change returns the total number of fields at the segment level, allowing for a more accurate estimate of the memory used by Lucene. The new estimate is expected to be closer to the actual memory usage than the current estimate using the index-level field count, due to the non-trivial overhead incurred by each Lucene segment. Two new fields are introduced: total_segment_fields, which represents the total number of fields at the segment level, and average_fields_per_segment. The overhead per field in segments with fewer fields is larger than in segments with many fields.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/NodeMappingStats.java [hunk 1]
```diff
@@ -63,17 +76,31 @@ public ByteSizeValue getTotalEstimatedOverhead() {
...
+        }
     }
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(Fields.MAPPINGS);
         builder.field(Fields.TOTAL_COUNT, getTotalCount());
         builder.humanReadableField(Fields.TOTAL_ESTIMATED_OVERHEAD_IN_BYTES, Fields.TOTAL_ESTIMATED_OVERHEAD, getTotalEstimatedOverhead());
+        builder.field(Fields.TOTAL_SEGMENT_FIELDS, totalSegments);
+        builder.field(Fields.AVERAGE_FIELDS_PER_SEGMENT, totalSegments == 0 ? 0 : totalSegmentFields / totalSegments);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/index/mapper/NodeMappingStats.java:103]: Having the average per node is nice but could we also include the raw counts in the output so we can aggregate this across nodes more easily?

File: server/src/main/java/org/elasticsearch/index/mapper/NodeMappingStats.java [hunk 2]
```diff
@@ -63,17 +76,31 @@ public ByteSizeValue getTotalEstimatedOverhead() {
...
+            out.writeVLong(totalSegmentFields);
+        }
     }
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(Fields.MAPPINGS);
         builder.field(Fields.TOTAL_COUNT, getTotalCount());
         builder.humanReadableField(Fields.TOTAL_ESTIMATED_OVERHEAD_IN_BYTES, Fields.TOTAL_ESTIMATED_OVERHEAD, getTotalEstimatedOverhead());
+        builder.field(Fields.TOTAL_SEGMENT_FIELDS, totalSegments);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/index/mapper/NodeMappingStats.java:102]: The field name suggests this should be the following perhaps?

```suggestion
        builder.field(Fields.TOTAL_SEGMENT_FIELDS, totalSegmentFields);
```

**Review by DaveCTurner — COMMENTED** [review]
Looking forward to having these stats available! I left a couple of comments.

File: server/src/main/java/org/elasticsearch/index/mapper/NodeMappingStats.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/mapper/NodeMappingStats.java]: Thanks David. Yes, we send these raw stats between nodes and return only the average in the rest response, as including the number of segments in the mappings section doesn't make sense.

File: server/src/main/java/org/elasticsearch/index/mapper/NodeMappingStats.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/mapper/NodeMappingStats.java]: Good catch!

File: server/src/main/java/org/elasticsearch/index/mapper/NodeMappingStats.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/mapper/NodeMappingStats.java]: > including the number of segments in the mappings section doesn't make sense

Although I see what you mean, I disagree that this is a good reason to omit it. The number of segments might not be available elsewhere (hence why we've added it to `NodeMappingStats`) or at least there may be some variance between the number used in this computation and the one exposed elsewhere. Let's have the raw figures here as well as the average please.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Overall change looks ok but I'd rather see the raw values in the output as well as the computed average (see [this comment](https://github.com/elastic/elasticsearch/pull/111123#discussion_r1687520907)).

**Review by andreidan — APPROVED** [review]
Thanks for working on this Nhat. Very nice to have these stats available ! 

Should we also update the docs (`node-stats.asciidoc`) ? 

LGTM, pending David's suggestion to expose the raw values as well

File: server/src/main/java/org/elasticsearch/index/mapper/NodeMappingStats.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/mapper/NodeMappingStats.java]: Sure, I pushed https://github.com/elastic/elasticsearch/pull/111123/commits/e42a69fb607110e9c7290e63f8e680ed93b1c0f1

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
@DaveCTurner @andreidan Thanks for reviewing :).

**Comment by dnhatn** [comment]
> Should we also update the docs (node-stats.asciidoc) ?

@andreidan ++ I updated it in https://github.com/elastic/elasticsearch/pull/111123/commits/9eb89c9bb32d97b72696f2ba01fa4f28f590db1a.

**Pull request merged**